karapace/auth.py

```diff
@@ -0,0 +1,185 @@
+from base64 import b64encode
+from dataclasses import dataclass, field
+from enum import Enum, unique
+from hmac import compare_digest
+from karapace.config import InvalidConfiguration
+from karapace.rapu import JSON_CONTENT_TYPE
+from typing import Optional
+
+import aiohttp
+import aiohttp.web
+import argparse
+import asyncio
+import base64
+import hashlib
+import json
+import logging
+import os
+import re
+import secrets
+import sys
+
+log = logging.getLogger(__name__)
+
+
+@unique
+class Operation(Enum):
+    Read = "Read"
+    Write = "Write"
+
+
+def hash_password(algorithm: str, salt: str, password: str) -> str:
+    if algorithm in ["sha1", "sha256", "sha512"]:
+        return b64encode(
+            hashlib.pbkdf2_hmac(algorithm, bytearray(password, "UTF-8"), bytearray(salt, "UTF-8"), 5000)
+        ).decode("ascii")
+    if algorithm == "scrypt":
+        return str(
+            base64.b64encode(hashlib.scrypt(bytearray(password, "utf-8"), salt=bytearray(salt, "utf-8"), n=16384, r=8, p=1)),
+            encoding="utf-8",
+        )
+    raise NotImplementedError(f"Hash algorithm '{algorithm}' is not implemented")
+
+
+@dataclass
+class User:
+    username: str
+    algorithm: str
+    salt: str
+    password: str = field(repr=False)  # hashed
+
+    def compare_password(self, plaintext: str) -> bool:
+        return compare_digest(self.password, hash_password(self.algorithm, self.salt, plaintext))
+
+
+@dataclass(frozen=True)
+class ACL:
+    username: str
+    operation: Operation
+    resource: re.Pattern
+
+
+class HTTPAuthorizer:
+    def __init__(self, filename: str):
+        self._auth_filename: str = filename
+        self._refresh_auth_task: Optional[asyncio.Task] = None
+        # Once first, can raise if file not valid
+        self._load_authfile()
+
+    async def start_refresh_task(self, app: aiohttp.web.Application) -> None:  # pylint: disable=unused-argument
+        async def _refresh_authfile() -> None:
+            """Reload authfile, but keep old auth data if loading fails"""
+
+            last_loaded = os.path.getmtime(self._auth_filename)
+
+            while True:
+                try:
+                    await asyncio.sleep(5)
+                    last_modified = os.path.getmtime(self._auth_filename)
+                    if last_loaded < last_modified:
+                        self._load_authfile()
+                        last_loaded = last_modified
```

Comment:
It is true that this causes `FATAL` log message every 5 seconds if the file is not valid. However logging `FATAL` once might go unnoticed too easily, and is this causes old users & acl rules to be in use after trying to change those, it could be justified to keep logging this continuously... but you raised a good point to think still

Reply:
It is true that this causes `FATAL` log message every 5 seconds if the file is not valid. However logging `FATAL` once might go unnoticed too easily, and is this causes old users & acl rules to be in use after trying to change those, it could be justified to keep logging this continuously... but you raised a good point to think still

Reply:
This does not get reported anywhere as the `Exception` is catched and not re-raised.
See: https://github.com/aiven/karapace/blob/36575e34230ce7a4c6d44e452d07cfe88a9c1cf9/karapace/rapu.py#L384

Only way to notice this from user perspective is that authentication fails for new credentials. Then the logs need to be reviewed.

I'd say the reattempt is ok and the logging is ok but the error handling scenario is not complete.